File: le-serviceflows/scoring/src/main/java/com/latticeengines/scoring/dataflow/ev/PercentileCalculationHelper.java
package com.latticeengines.scoring.dataflow.ev;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import javax.inject.Inject;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

import com.latticeengines.dataflow.exposed.builder.Node;
import com.latticeengines.dataflow.exposed.builder.common.FieldList;
import com.latticeengines.dataflow.runtime.cascading.cdl.CalculatePercentile;
import com.latticeengines.domain.exposed.camille.CustomerSpace;
import com.latticeengines.domain.exposed.dataflow.FieldMetadata;
import com.latticeengines.domain.exposed.scoring.ScoreResultField;
import com.latticeengines.scoring.dataflow.CalculateExpectedRevenuePercentile.ParsedContext;

import cascading.tuple.Fields;

@Component("percentileCalculationHelper")
public class PercentileCalculationHelper {

    private static final Logger log = LoggerFactory.getLogger(PercentileCalculationHelper.class);

    @Inject
    private NodeSplitter nodeSplitter;

    public Node calculate(ParsedContext context, Node mergedScoreCount, boolean trySecondarySort,
            boolean targetScoreDerivation) {
        Node[] nodes = nodeSplitter.splitEv(mergedScoreCount, context.originalScoreFieldMap,
                context.modelGuidFieldName);
        String secondarySortFieldName = trySecondarySort ? context.outputExpRevFieldName : null;
        Node model = getModel(context, nodes[0], secondarySortFieldName, ScoreResultField.RawScore.displayName,
                targetScoreDerivation);
        Node predictedModel = getModel(context, nodes[1], secondarySortFieldName,
                ScoreResultField.PredictedRevenue.displayName, targetScoreDerivation);
        Node evModel = getModel(context, nodes[2], secondarySortFieldName, ScoreResultField.ExpectedRevenue.displayName,
                targetScoreDerivation);

        List<Node> models = Arrays.asList(model, predictedModel, evModel).stream().filter(e -> e != null)
                .collect(Collectors.toList());
        Node merged = models.get(0);
        for (int i = 1; i < models.size(); i++) {
            merged = merged.merge(models.get(i));
        }
        return merged;
    }

    private Node getModel(ParsedContext context, Node node, String secondarySortFieldName,
            String originalScoreFieldName, boolean targetScoreDerivation) {
        if (node == null) {
            return node;
        }
        return calculatePercentileByFieldName(context.modelGuidFieldName, context.scoreCountFieldName,
                originalScoreFieldName, context.percentileFieldName, context.standardScoreField, secondarySortFieldName,
                context.minPct, context.maxPct, node, context.targetScoreDerivationPaths, context.customerSpace,
                targetScoreDerivation);
    }

    private Node calculatePercentileByFieldName(String modelGuidFieldName, String scoreCountFieldName,
            String originalScoreFieldName, String percentileFieldName, String defaultPercentileFieldName,
            String secondarySortFieldName, int minPct, int maxPct, Node node,
            Map<String, String> targetScoreDerivationPaths, CustomerSpace customerSpace,
            boolean targetScoreDerivation) {
        if (ScoreResultField.RawScore.displayName.equals(originalScoreFieldName)) {
            return node;
        }

        node = node.addColumnWithFixedValue(percentileFieldName, null, Integer.class);
        List<String> returnedFields = new ArrayList<>(node.getFieldNames());
        List<FieldMetadata> returnedMetadata = new ArrayList<>(node.getSchema());
        FieldList sortFieldList = null;
        if (secondarySortFieldName == null || secondarySortFieldName.equals(originalScoreFieldName)) {
            sortFieldList = new FieldList(originalScoreFieldName);
        } else {
            sortFieldList = new FieldList(originalScoreFieldName, secondarySortFieldName);
        }

        log.info("Use target score derivation=" + targetScoreDerivation);
        Node calculatePercentile = node.groupByAndBuffer(new FieldList(modelGuidFieldName), sortFieldList,
                new CalculatePercentile(new Fields(returnedFields.toArray(new String[0])), minPct, maxPct,
                        percentileFieldName, scoreCountFieldName, originalScoreFieldName, targetScoreDerivation,
                        targetScoreDerivationPaths, modelGuidFieldName),
                true, returnedMetadata);
        return calculatePercentile;
    }
}

File: le-spark/src/main/scala/com/latticeengines/spark/exposed/job/score/PercentileCalculationHelper.scala
package com.latticeengines.spark.exposed.job.score

import com.latticeengines.domain.exposed.cdl.scoring.CalculatePercentile2
import com.latticeengines.domain.exposed.scoring.ScoreResultField
import org.apache.spark.sql.{ DataFrame }
import scala.collection.JavaConverters._
import org.apache.spark.sql.functions.{ col, udf, lit, desc, row_number, min, max }
import org.apache.spark.sql.types._
import scala.collection.mutable.{ Map => MMap }
import org.apache.spark.sql.expressions.Window
import scala.collection.mutable.ListBuffer
import collection.JavaConversions._
import org.apache.spark.storage.StorageLevel

object PercentileCalculationHelper {
  val INCREASE_ID = "__increasing_id"
  val MIN_COL = "__min_col"
  val MAX_COL = "__max_col"
  def calculate(context: ParsedContext, mergedScoreCount: DataFrame, trySecondarySort: Boolean,
    targetScoreDerivation: Boolean): DataFrame = {

    var (model, predictedModel, evModel) = NodeJobSplitter.splitEv(mergedScoreCount, context.originalScoreFieldMap,
            context.modelGuidFieldName)
            
    val secondarySortFieldName = if (trySecondarySort) context.outputExpRevFieldName else null
    model = getModel(context, model, secondarySortFieldName, ScoreResultField.RawScore.displayName,
            targetScoreDerivation)
    predictedModel = getModel(context, predictedModel, secondarySortFieldName,
            ScoreResultField.PredictedRevenue.displayName, targetScoreDerivation);
    evModel = getModel(context, evModel, secondarySortFieldName, ScoreResultField.ExpectedRevenue.displayName,
            targetScoreDerivation);
    
    val models = List(model, predictedModel, evModel).filter(e => e != null)
    var merged = models(0)
    for (i <- 1 until models.length) {
        merged = merged.union(models(i))
    }
    merged
        
  }

  def getModel(context: ParsedContext, node: DataFrame, secondarySortFieldName: String,
        originalScoreFieldName: String, targetScoreDerivation: Boolean): DataFrame = {
    if (node == null) {
      return node;
    }
    calculatePercentileByFieldName(context.modelGuidFieldName, context.scoreCountFieldName,
      originalScoreFieldName, context.percentileFieldName, context.standardScoreField, secondarySortFieldName,
      context.minPct, context.maxPct, node, context.targetScoreDerivationInputs, context.targetScoreDerivationOutputs,
      context.originalScoreFieldMap, targetScoreDerivation, true)
  }

  def calculatePercentileByFieldName(modelGuidFieldName: String, scoreCountFieldName: String,
    originalScoreFieldName: String, percentileFieldName: String, defaultPercentileFieldName: String,
    secondarySortFieldName: String, minPct: Integer, maxPct: Integer, nodeInput: DataFrame,
    targetScoreDerivationInputs: Map[String, String],
    targetScoreDerivationOutputs: MMap[String, String],
    originalScoreFieldMap: Map[String, String],
    targetScoreDerivation: Boolean, skipProbability: Boolean): DataFrame = {

    if ((ScoreResultField.RawScore.displayName == originalScoreFieldName) && skipProbability) {
      return nodeInput
    }

    println("----- BEGIN SCRIPT OUTPUT -----")
    println(s"modelGuid1=${modelGuidFieldName}")
    println("----- END SCRIPT OUTPUT -----")
    
    var node = nodeInput.withColumn(percentileFieldName, lit(null).cast(IntegerType))
    val returnedFields = List(node.columns)
    val sortFieldList = if (secondarySortFieldName == null || secondarySortFieldName == originalScoreFieldName) {
      List(originalScoreFieldName)
    } else {
      List(originalScoreFieldName, secondarySortFieldName)
    }

    val windowSpec = Window.partitionBy(modelGuidFieldName)
      .orderBy(sortFieldList map desc: _*)
    node = node.withColumn(INCREASE_ID, row_number().over(windowSpec))
    
    val calFunc = new CalculatePercentile2(minPct, maxPct,
        targetScoreDerivation,
        mapAsJavaMap(targetScoreDerivationInputs))
    
    val calFuncUdf = udf((modelId: String, rawScore: Double, totalCount: Long, increaseId: Long) => {
      calFunc.calculate(modelId, rawScore, totalCount, increaseId)
    })
    node = node.withColumn(percentileFieldName, calFuncUdf(
      col(modelGuidFieldName),
      col(originalScoreFieldName),
      col(scoreCountFieldName),
      col(INCREASE_ID) - 1))
      
    if (targetScoreDerivation) {
      val derivations = node.groupBy(modelGuidFieldName, percentileFieldName)
        .agg(min(col(originalScoreFieldName)).as(MIN_COL), max(col(originalScoreFieldName)).as(MAX_COL))
        .select(Seq(modelGuidFieldName, percentileFieldName, MIN_COL, MAX_COL) map col: _*).persist(StorageLevel.DISK_ONLY)
        
      for ((modelGuid, scoreField) <- originalScoreFieldMap) {
        if (!targetScoreDerivationInputs.contains(modelGuid)) {
          println("----- BEGIN SCRIPT OUTPUT -----")
          println(s"modelGuid5=${modelGuidFieldName}")
          println("----- END SCRIPT OUTPUT -----")
          
          val derivationMap = new java.util.HashMap[Integer, java.util.List[java.lang.Double]]()
          var derivation = derivations.filter(col(modelGuidFieldName) === modelGuid)
          derivation.collect().foreach(row => {
            val minMax = new java.util.ArrayList[java.lang.Double]()
            val pct = row.getInt(1)
            val minV = row.getDouble(2)
            val maxV = row.getDouble(3)
            minMax.add(minV)
            minMax.add(maxV)
            derivationMap.put(pct, minMax)
          })
          val targetScoreDerivationStr = calFunc.getTargetScoreDerivationValue(modelGuid, derivationMap)
          if (targetScoreDerivationStr != null) {
            targetScoreDerivationOutputs(modelGuid) = targetScoreDerivationStr
          }
        }
      }
    }
    node.drop(INCREASE_ID)
  }
}
